src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5127,100 +5128,113 @@ private void startLeaving()
 
     public void decommission(boolean force) throws InterruptedException
     {
-        TokenMetadata metadata = tokenMetadata.cloneAfterAllLeft();
-        if (operationMode != Mode.LEAVING)
+        try
         {
-            if (!tokenMetadata.isMember(FBUtilities.getBroadcastAddressAndPort()))
-                throw new UnsupportedOperationException("local node is not a member of the token ring yet");
-            if (metadata.getAllEndpoints().size() < 2)
+            TokenMetadata metadata = tokenMetadata.cloneAfterAllLeft();
+            if (operationMode != Mode.LEAVING)
+            {
+                if (!tokenMetadata.isMember(FBUtilities.getBroadcastAddressAndPort()))
+                    throw new UnsupportedOperationException("local node is not a member of the token ring yet");
+                if (metadata.getAllEndpoints().size() < 2)
                     throw new UnsupportedOperationException("no other normal nodes in the ring; decommission would be pointless");
-            if (operationMode != Mode.NORMAL)
-                throw new UnsupportedOperationException("Node in " + operationMode + " state; wait for status to become normal or restart");
-        }
-        if (!isDecommissioning.compareAndSet(false, true))
-            throw new IllegalStateException("Node is still decommissioning. Check nodetool netstats.");
+                if (operationMode != Mode.NORMAL)
+                    throw new UnsupportedOperationException("Node in " + operationMode + " state; wait for status to become normal or restart");
+            }
+            if (!isDecommissioning.compareAndSet(false, true))
```

Comment:
I think we can not include some code into the try catch , like line 5143 
` if (!isDecommissioning.compareAndSet(false, true))`
first time you have execute a decommission, and the node is being doing decommission which may take a long time ,
and after a while you may execute another decommission , you will got a false. But actually the node's decommission is be doing now and have not failed. 
you will get a wrong status.